File: c/src/_0278/firstBadVersion.c
//
// Created by gaojian on 2022/8/12.
//

#include "firstBadVersion.h"
//
//int firstBadVersion(int n) {
//    int left = 0;
//    int mid = 0;
//
//    while (left < n) {
//        mid = left + (n - left) / 2 ;
//       if (isBadVersion(mid)) {
//           n = mid;
//       } else {
//           left = mid + 1;
//       }
//    }
//
//    return left;
//}

int firstBadVersion(int n) {
    long long first = 1, last = (long long ) n + 1;
    long long mid = first + (last - first) / 2;
    while (isBadVersion(mid) && !isBadVersion(mid - 1)) {
        mid = first + (last - first) / 2;
        if (!isBadVersion(mid)) {
            first = mid;
        } else {
            last = mid;
        }
    }

    return 0;
}

int binarySearch(long long left, long long right) {
    long long mid = left + (left - right) / 2;

    if (isBadVersion(mid) && isBadVersion(mid - 1)) {
      return mid;
    }

    if (!isBadVersion(mid)) {
        return binarySearch(mid + 1, right);
    }

    return binarySearch(left, mid);
}

File: go/src/easy/leetcode-278/firstBadVersion.go
package leetcode278

/**
 * Forward declaration of isBadVersion API.
 * @param   version   your guess about first bad version
 * @return 	 	      true if current version is bad
 *			          false if current version is good
 * func isBadVersion(version int) bool;
 */

func defaultSearch(n int, bad int) int {
	for i := 1; i <= n; i++ {
		if isBadVersion(i, bad) {
			return i
		}
	}

	return 0
}

func binarySearch(n int, bad int) int {
	l := 0
	for l < n {
		mid := l + (n-l)/2
		if isBadVersion(mid, bad) {
			n = mid
		} else {
			l = mid + 1
		}
	}

	return l
}

func isBadVersion(n int, bad int) bool {
	return n == bad
}
